File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
def check(mid,v,m,k):
    n=len(v)
    vis=[0]*n
    for i in range(n):
        if mid>=v[i]:
            vis[i]=1

    adj=0
    ans=0
    for i in range(n):
        if vis[i]==1:
            adj+=1
        else:
            if adj>=k:
                m-=1
            adj=0
        
        if adj>=k:
            m-=1
            adj=0
    
    return m<=0


class Solution:
    def minDays(self, v: List[int], m: int, k: int) -> int:
        l=min(v)
        h=max(v)
        ans=h
        while l<=h:
            mid=(l+h)//2
            if check(mid,v,m,k):
                ans=min(ans,mid)
                h=mid-1
            else:
                l=mid+1
        if ans==h:
            return -1
        return ans

        

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
bool can_make(int day,vector<int>&arr,int adj,int need){
    int n=arr.size();
    int cnt=0;
    int no_bouquets=0;
    for(int i=0;i<n;i++){

        if(arr[i]<=day){
            cnt++;
        }
        else{
            no_bouquets+=cnt/adj;
            cnt=0;
        }
    }
    no_bouquets+=cnt/adj;
    return no_bouquets>=need;
}


class Solution {
public:
    int minDays(vector<int>& arr, int m, int k) {

        int l=1;
        int h=*max_element(arr.begin(),arr.end());

        int ans=-1;

        while(l<=h){
            int mid=(l+h)/2;
            if(can_make(mid,arr,k,m)){
                ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }
        }
        return ans;
        
    }
};